clusterloader2/pkg/measurement/common/wait_for_controlled_pods.go

```diff
@@ -24,123 +24,385 @@ package common
+		w.workerGroup.Start(w.worker)
+	}
+	go w.informer.Run(w.stopCh)
+	timeoutCh := make(chan struct{})
+	timeoutTimer := time.AfterFunc(informerSyncTimeout, func() {
+		close(timeoutCh)
+	})
+	defer timeoutTimer.Stop()
+	if !cache.WaitForCacheSync(timeoutCh, w.informer.HasSynced) {
+		return fmt.Errorf("timed out waiting for caches to sync")
+	}
+
+	return nil
+}
+
+func (w *waitForControlledPodsRunningMeasurement) worker() {
+	for {
+		f, stop := w.queue.Get()
+		if stop {
+			return
+		}
+		(*f.(*func()))()
+		w.queue.Done(f)
+	}
+}
+
+func (w *waitForControlledPodsRunningMeasurement) gather(c clientset.Interface, syncTimeout time.Duration) error {
+	glog.Infof("%v: waiting for controlled pods measurement...", w)
+	if w.informer == nil {
+		return fmt.Errorf("metric %s has not been started", w)
+	}
+	desiredCount, maxResourceVersion, err := w.getObjectCountAndMaxVersion(c)
 	if err != nil {
-		return summaries, err
+		return err
+	}
+
+	if err := wait.Poll(checkControlledPodsInterval, syncTimeout, func() (bool, error) { return w.opResourceVersion >= maxResourceVersion, nil }); err != nil {
+		return fmt.Errorf("timed out while waiting for controlled pods")
+	}
+
+	w.handlingGroup.Wait()
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	numberRunning := 0
+	for _, checker := range w.checkerMap {
+		if status := checker.getStatus(); status {
+			numberRunning++
+		}
+	}
+	if desiredCount != numberRunning {
+		glog.Errorf("%s: incorrect objects number: %d/%d %ss are running with all pods", w, numberRunning, desiredCount, w.kind)
+		return fmt.Errorf("incorrect objects number")
+	}
+
+	glog.Infof("%s: %d/%d %ss are running with all pods", w, numberRunning, desiredCount, w.kind)
+	return nil
+}
+
+func (w *waitForControlledPodsRunningMeasurement) stop() {
+	close(w.stopCh)
+	w.queue.ShutDown()
+	w.workerGroup.Wait()
+	w.lock.Lock()
+	defer w.lock.Unlock()
+	w.isRunning = false
+	for _, checker := range w.checkerMap {
+		checker.terminate()
+	}
+}
+
+func (w *waitForControlledPodsRunningMeasurement) handleObject(c clientset.Interface, oldObj, newObj interface{}) {
+	var oldRuntimeObj runtime.Object
+	var newRuntimeObj runtime.Object
+	var ok bool
+	oldRuntimeObj, ok = oldObj.(runtime.Object)
+	if oldObj != nil && !ok {
+		return
+	}
+	defer func() {
+		// We want to update version after (potentially) creating goroutine.
+		if err := w.updateOpResourceVersion(oldRuntimeObj); err != nil {
+			glog.Errorf("%s: updating resource version error: %v", w, err)
+		}
+	}()
+	newRuntimeObj, ok = newObj.(runtime.Object)
+	if newRuntimeObj != nil && !ok {
+		return
+	}
+	defer func() {
+		// We want to update version after (potentially) creating goroutine.
+		if err := w.updateOpResourceVersion(newRuntimeObj); err != nil {
+			glog.Errorf("%s: updating resource version error: %v", w, err)
+		}
+	}()
+	if isEqual, err := runtimeobjects.IsEqualRuntimeObjectsSpec(oldRuntimeObj, newRuntimeObj); err != nil || isEqual {
+		if err != nil {
```

Comment:
We generally should try to avoid nesting whenever it's possible. So this is better.
```
isEqual, err := ...
if err != nil {
  glog.Errorf("...")
  return
}
if isEqual {
  // Add comment
  return
}
```

Even though it's a bit longer, the flow is much more obvious now.

Reply:
Done.